ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs

```diff
@@ -0,0 +1,306 @@
+{-# LANGUAGE DeriveFunctor         #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+
+-- | Error policies, and integration with 'SuspendDecision'-semigroup action on
+-- 'PeerState'.
+--
+module Ouroboros.Network.ErrorPolicy
+  ( ErrorPolicies (..)
+  , nullErrorPolicies
+  , ErrorPolicy (..)
+  , evalErrorPolicy
+  , evalErrorPolicies
+  , ConnectionOrApplicationException (..)
+  , CompleteApplication
+  , Result (..)
+  , completeApplicationTx
+
+  -- * Traces
+  , ErrorPolicyTrace (..)
+  , traceErrorPolicy
+  , WithAddr (..)
+
+  -- * Re-exports of PeerState
+  , PeerStates
+  , SuspendDecision (..)
+  ) where
+
+import           Control.Exception (Exception, SomeException (..))
+import           Data.List.NonEmpty (NonEmpty (..))
+import           Data.Foldable (traverse_)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (fromMaybe, mapMaybe)
+import           Data.Semigroup (sconcat)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Typeable ( Proxy (..)
+                               , gcast
+                               , tyConName
+                               , typeRepTyCon
+                               , typeRep
+                               )
+import           Text.Printf
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Tracer
+
+import           Data.Semigroup.Action
+
+import           Ouroboros.Network.Subscription.PeerState
+
+data ErrorPolicy where
+     ErrorPolicy :: forall err.
+                      Exception err
+                   => (ConnectionOrApplicationException err -> SuspendDecision DiffTime)
```

Comment:
Should the consensus policies inspect which constructor of `ConnectionOrApplicationException` is used? I'm assuming it is mainly of interest to the handler of the policies.

Reply:
Connection exceptions policy should be supplied by `ouroboros-network` (one can only expect an `IOException` there),  while application exceptions policies should be supplied by `ouroboros-consensus`. I think we should have a clearer split between those two cases.